In [130]:
import pandas as pd
import math
from functools import reduce
import numpy as np
from statsmodels import robust
import scipy


df_idle = pd.DataFrame()

for i in range(1,1040):
    df_win = pd.read_csv(f"data\idle\idle-{i}.csv")
    
    df_extr = pd.DataFrame()
    df_extr['x_max'] = {i:df_win.iloc[:,0].max()}
    df_extr['x_min'] = [df_win.iloc[:,0].min()]
    df_extr['x_ent'] = -reduce(
        lambda x,y:x+y,
        map(
            lambda x:(x/len(df_win))*math.log2(x/len(df_win)),
            df_win.iloc[:,0].value_counts()))
    q75, q25 = np.percentile(df_win.iloc[:,0], [75 ,25])
    df_extr['x_int'] = [q75 - q25]
    df_extr['y_max'] = [df_win.iloc[:,1].max()]
    df_extr['y_xmin'] = [np.where(df_win==df_win.iloc[:, 1].min())[0][0]]
    df_extr['y_mdiv'] = [robust.mad(df1.iloc[:,1])]
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_skew'] = scipy.stats.skew(df_win.iloc[:,1], bias= False)
    df_extr['y_std'] = df_win.iloc[:,1].std()
    df_extr['z_skew'] = scipy.stats.skew(df_win.iloc[:,2], bias= False)
    df_extr['Y'] = 1
    df_idle = pd.concat([df_idle, df_extr])

print(df_idle.shape)
df_idle.head()

(1039, 12)


x_max     x_min     x_ent     x_int     y_max  y_xmin    y_mdiv  \
1  5.099650 -0.909797  4.523231  0.065840  4.616021       2  0.031947   
2  5.099650 -0.909797  4.389898  0.049081  0.148441       0  0.031947   
3  1.762132 -0.320823  4.281728  0.040701  0.023942      21  0.031947   
4  0.407014 -0.320823  4.069577  0.037110  0.023942      19  0.031947   
5  0.407014 -0.320823  3.939572  0.035913  0.023942      17  0.031947   

      y_med    y_skew     y_std    z_skew  Y  
1 -0.131681  3.647470  1.157603 -2.365312  1  
2 -0.134075  1.839191  0.075092 -3.844882  1  
3 -0.131681  0.820672  0.045866 -4.985304  1  
4 -0.129287  0.712333  0.045738 -2.839063  1  
5 -0.129287  0.664670  0.046731 -2.452804  1

In [131]:
df_running = pd.DataFrame()

for i in range(1,3409):
    df_win = pd.read_csv(f"data\\running\\running-{i}.csv")
    
    df_extr = pd.DataFrame()
    df_extr['x_max'] = {i:df_win.iloc[:,0].max()}
    df_extr['x_min'] = [df_win.iloc[:,0].min()]
    df_extr['x_ent'] = -reduce(
        lambda x,y:x+y,
        map(
            lambda x:(x/len(df_win))*math.log2(x/len(df_win)),
            df_win.iloc[:,0].value_counts()))
    q75, q25 = np.percentile(df_win.iloc[:,0], [75 ,25])
    df_extr['x_int'] = [q75 - q25]
    df_extr['y_max'] = [df_win.iloc[:,1].max()]
    df_extr['y_xmin'] = [np.where(df_win==df_win.iloc[:, 1].min())[0][0]]
    df_extr['y_mdiv'] = [robust.mad(df1.iloc[:,1])]
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_skew'] = scipy.stats.skew(df_win.iloc[:,1], bias= False)
    df_extr['y_std'] = df_win.iloc[:,1].std()
    df_extr['z_skew'] = scipy.stats.skew(df_win.iloc[:,2], bias= False)
    df_extr['Y'] = 2
    df_running = pd.concat([df_running, df_extr])

print(df_running.shape)
df_running.head()

(3408, 12)


x_max      x_min     x_ent      x_int      y_max  y_xmin    y_mdiv  \
1  22.720974  -4.941632  4.906891   6.890514  19.091364      27  0.031947   
2  22.720974  -4.941632  4.906891   9.471462  19.091364      25  0.031947   
3  25.914839  -7.220912  4.906891  11.243172  19.091364      23  0.031947   
4  25.914839  -7.220912  4.906891  13.085510  19.091364      21  0.031947   
5  25.914839 -10.994174  4.906891  13.085510  19.091364      28  0.031947   

      y_med    y_skew     y_std    z_skew  Y  
1  6.210560 -0.685412  8.509691  0.860710  2  
2  7.374142 -0.832846  8.601343  0.970486  2  
3  7.374142 -0.766092  8.603497  1.014803  2  
4  7.745242 -0.827988  8.665222  1.064640  2  
5  8.104373 -1.134891  9.857067  1.061706  2

In [132]:
df_stairs = pd.DataFrame()

for i in range(1,166):
    df_win = pd.read_csv(f"data\stairs\stairs-{i}.csv")
    
    df_extr = pd.DataFrame()
    df_extr['x_max'] = {i:df_win.iloc[:,0].max()}
    df_extr['x_min'] = [df_win.iloc[:,0].min()]
    df_extr['x_ent'] = -reduce(
        lambda x,y:x+y,
        map(
            lambda x:(x/len(df_win))*math.log2(x/len(df_win)),
            df_win.iloc[:,0].value_counts()))
    q75, q25 = np.percentile(df_win.iloc[:,0], [75 ,25])
    df_extr['x_int'] = [q75 - q25]
    df_extr['y_max'] = [df_win.iloc[:,1].max()]
    df_extr['y_xmin'] = [np.where(df_win==df_win.iloc[:, 1].min())[0][0]]
    df_extr['y_mdiv'] = [robust.mad(df1.iloc[:,1])]
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_skew'] = scipy.stats.skew(df_win.iloc[:,1], bias= False)
    df_extr['y_std'] = df_win.iloc[:,1].std()
    df_extr['z_skew'] = scipy.stats.skew(df_win.iloc[:,2], bias= False)
    df_extr['Y'] = 3    
    df_stairs = pd.concat([df_stairs, df_extr])

print(df_stairs.shape)
df_stairs.head()

(165, 12)


x_max     x_min     x_ent     x_int     y_max  y_xmin    y_mdiv  \
1  8.073248 -4.582502  4.906891  3.848680  2.973599      19  0.031947   
2  8.073248 -4.582502  4.906891  3.539828 -0.483629      17  0.031947   
3  8.288726 -4.582502  4.906891  3.529054 -0.483629      15  0.031947   
4  8.288726 -4.582502  4.906891  3.629610 -0.483629      28  0.031947   
5  8.288726 -4.582502  4.906891  3.824737 -0.483629      26  0.031947   

      y_med    y_skew     y_std    z_skew  Y  
1 -7.153874 -0.280692  4.710145 -0.473734  3  
2 -7.414842 -0.764308  4.037996 -0.579582  3  
3 -7.596801 -0.780079  3.847049 -0.546646  3  
4 -7.963115 -0.872657  4.249331 -0.646571  3  
5 -7.963115 -0.869825  4.230499 -0.598127  3

In [133]:
df_walking = pd.DataFrame()

for i in range(1,1851):
    df_win = pd.read_csv(f"data\walking\walking-{i}.csv")
    
    df_extr = pd.DataFrame()
    df_extr['x_max'] = {i:df_win.iloc[:,0].max()}
    df_extr['x_min'] = [df_win.iloc[:,0].min()]
    df_extr['x_ent'] = -reduce(
        lambda x,y:x+y,
        map(
            lambda x:(x/len(df_win))*math.log2(x/len(df_win)),
            df_win.iloc[:,0].value_counts()))
    q75, q25 = np.percentile(df_win.iloc[:,0], [75 ,25])
    df_extr['x_int'] = [q75 - q25]
    df_extr['y_max'] = [df_win.iloc[:,1].max()]
    df_extr['y_xmin'] = [np.where(df_win==df_win.iloc[:, 1].min())[0][0]]
    df_extr['y_mdiv'] = [robust.mad(df1.iloc[:,1])]
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_med'] = df_win.iloc[:,1].median()
    df_extr['y_skew'] = scipy.stats.skew(df_win.iloc[:,1], bias= False)
    df_extr['y_std'] = df_win.iloc[:,1].std()
    df_extr['z_skew'] = scipy.stats.skew(df_win.iloc[:,2], bias= False)
    df_extr['Y'] = 4    
    df_walking = pd.concat([df_walking, df_extr])

print(df_walking.shape)
df_walking.head()

(1850, 12)


x_max     x_min     x_ent     x_int     y_max  y_xmin    y_mdiv  \
1  12.363658 -2.389413  4.906891  4.782418  4.668694       8  0.031947   
2  12.363658 -2.389413  4.906891  5.373786 -1.398214       6  0.031947   
3  12.363658 -2.389413  4.906891  5.373786 -2.159570       4  0.031947   
4  12.363658 -2.389413  4.906891  5.363012 -2.159570       2  0.031947   
5  12.363658 -2.964022  4.906891  5.331888 -2.159570       0  0.031947   

       y_med    y_skew     y_std    z_skew  Y  
1 -10.472238  0.659964  5.590863 -1.635672  4  
2 -10.654198 -0.063457  4.490179 -0.171282  4  
3 -10.826580 -0.236329  4.355248 -0.283234  4  
4 -10.716447 -0.204809  4.627718 -0.360408  4  
5 -10.500969 -0.279222  4.788349 -0.390847  4

In [134]:
df = pd.concat([df_idle, df_running, df_stairs, df_walking], ignore_index=True)
df.shape

(6462, 12)

In [304]:
from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC, LinearSVC


df_x = df.drop(["Y"], axis=1)
df_y = df[['Y']].to_numpy()

X_train, X_test, y_train, y_test = train_test_split(
    df_x, df_y, test_size=0.33, shuffle=True, random_state=0)
clf = SVC(decision_function_shape='ovo', kernel="rbf").fit(X_train, y_train.ravel())

clf.score(X_test, y_test)

0.9718706047819972

In [302]:
clf = SVC(decision_function_shape='ovo', kernel='linear').fit(X_train, y_train.ravel())

clf.score(X_test, y_test)

0.9845288326300985

In [305]:
clf = SVC(decision_function_shape='ovo', kernel='poly').fit(X_train, y_train.ravel())

clf.score(X_test, y_test)

0.9695264885138303

In [303]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier 
 

tree_clf = DecisionTreeClassifier(max_depth=10) 
tree_clf.fit(X_train, y_train)

tree_clf.score(X_test, y_test)

0.9967182372245663

 У роботі підготовлено датасет і розраховано додаткові часові ознаки (time domain features).
 Проаналізовано роботу алгоритмів Методу опорних векторів з лінійною, поліноміальною, радіальною базисною функціями та Методу дерев рішень. Оцінка якості роботи алгорітмів була проведена за допомогою метрики accuracy. 
 По результатам роботи можна зробити такі висновки, що за допомогою розрахованих часових ознак можна досягти високої точності роботи класифікаторів. Всі класифікатори дають досить високі показники показники, проте, найменшу точність показує Метод опорних векторів з поліноміальним ядром (точність = 0.9695), а найбільшу точність показує Метод дерева рішень (точність = 0.9967)